http2_general/test_h2_frame.py

```diff
@@ -46,20 +49,203 @@ def test_empty_last_data_frame(self):
+
         deproxy_cl.make_request(request=self.post_request, end_stream=False)
         deproxy_cl.send_bytes(DataFrame(stream_id=1, data=b"").serialize())
-        deproxy_cl.make_request(request=request_body, end_stream=True)
 
-        self.__assert_test(client=deproxy_cl, request_body=request_body, request_number=1)
+        self.assertTrue(deproxy_cl.wait_for_connection_close(timeout=5))
+        self.assertIn(ErrorCodes.PROTOCOL_ERROR, deproxy_cl.error_codes)
+
+    def test_multiple_empty_headers_frames(self):
+        """
+        An endpoint that receives a HEADERS frame without the END_STREAM flag set
+        after receiving the HEADERS frame that opens a request or after receiving
+        a final (non-informational) status code MUST treat the corresponding
+        request or response as malformed (Section 8.1.1).
+        """
+        self.start_all_services()
+        deproxy_cl = self.get_client("deproxy")
+        deproxy_cl.parsing = False
+
+        self.initiate_h2_connection(deproxy_cl)
+        # create stream and change state machine in H2Connection object
+        stream = deproxy_cl.h2_connection._get_or_create_stream(
+            deproxy_cl.stream_id, AllowedStreamIDs(deproxy_cl.h2_connection.config.client_side)
+        )
+        stream.state_machine.process_input(StreamInputs.SEND_HEADERS)
+
+        hf_empty = HeadersFrame(
+            stream_id=1,
+        )
+
+        hf_end = HeadersFrame(
+            stream_id=1,
+            data=deproxy_cl.h2_connection.encoder.encode(self.get_request),
+            flags=["END_HEADERS", "END_STREAM"],
+        )
+
+        for i in range(20):
+            deproxy_cl.send_bytes(hf_empty.serialize())
+        deproxy_cl.send_bytes(hf_end.serialize(), expect_response=True)
+
+        self.assertTrue(deproxy_cl.wait_for_connection_close(timeout=5))
+        self.assertIn(ErrorCodes.PROTOCOL_ERROR, deproxy_cl.error_codes)
+
+    @parameterize.expand(
+        [
+            param(name="end_headers", end_headers=True),
+            param(name="no_end_headers", end_headers=False),
+        ]
+    )
+    def test_empty_headers_frame(self, name, end_headers):
+        """
+        An endpoint that receives a HEADERS frame without the END_STREAM flag set
+        after receiving the HEADERS frame that opens a request or after receiving
+        a final (non-informational) status code MUST treat the corresponding
+        request or response as malformed (Section 8.1.1).
+        """
+        self.start_all_services()
+        deproxy_cl = self.get_client("deproxy")
+        deproxy_cl.parsing = False
+
+        self.initiate_h2_connection(deproxy_cl)
+        # create stream and change state machine in H2Connection object
+        stream = deproxy_cl.h2_connection._get_or_create_stream(
+            deproxy_cl.stream_id, AllowedStreamIDs(deproxy_cl.h2_connection.config.client_side)
+        )
+        stream.state_machine.process_input(StreamInputs.SEND_HEADERS)
```

Comment:
Maybe we should to add this code as new method to DeproxyClientH2?

Reply:
Done.